go/pkg/cosmos/txhistory.go

```diff
@@ -126,92 +162,416 @@ func Messages(msgs []sdk.Msg) []Message {
+			if txHeight > s.cursor.LastBlockHeight {
+				continue
+			}
+			// if equal to last height from cursor, skip until we find matching last tx
+			if txHeight == s.cursor.LastBlockHeight {
+				if *tx.Hash == lastTxID {
+					foundLastTx = true
+					// found as last tx in results, need to fetch next page
+					if i == len(res.Result.Txs)-1 {
+						refetch = true
+					}
+					continue
+				}
+				if !foundLastTx {
+					continue
+				}
+			}
+
+			filtered = append(filtered, tx)
+		}
+		if !refetch {
+			res.Result.Txs = filtered
+			break
+		}
+		refetch = false
+		page++
+	}
+
+	switch txType {
+	case histTxTypeSender:
+		s.cursor.TendermintSentPage = page
+	case histTxTypeRecipient:
+		s.cursor.TendermintRecvdPage = page
+	default:
+		return nil, fmt.Errorf("invalid histTxType %d", txType)
 	}
 
-	totalSend, err := strconv.ParseFloat(res.Result.TotalCount, 64)
+	return &res, nil
+}
+
+func (c *txHistoryState) nextSenderTxHistoryPage() (*client.TxSearchResponse, error) {
+	c.cursor.TendermintSentPage++
+	res, err := c.tendermintHistoryRequest(histTxTypeSender)
 	if err != nil {
-		return nil, errors.Wrapf(err, "failed to parse totalSend: %s", res.Result.TotalCount)
+		return nil, errors.Wrapf(err, "error getting tendermint sent tx history")
 	}
 
-	resTxs := res.Result.Txs
+	return res, nil
+}
 
-	res, _, err = c.tendermintClient.InfoApi.TxSearch(c.ctx).Query(fmt.Sprintf("\"transfer.recipient='%s'\"", address)).Page(int32(page)).PerPage(int32(pageSize)).OrderBy("\"desc\"").Execute()
+func (c *txHistoryState) nextRecipTxHistoryPage() (*client.TxSearchResponse, error) {
+	c.cursor.TendermintRecvdPage++
+	res, err := c.tendermintHistoryRequest(histTxTypeRecipient)
 	if err != nil {
-		return nil, errors.Wrap(err, "failed to get receive transactions")
+		return nil, errors.Wrapf(err, "error getting tendermint recipient tx history")
 	}
 
-	totalReceive, err := strconv.ParseFloat(res.Result.TotalCount, 64)
+	return res, nil
+}
+
+func buildSentTxsQuery(address string) string {
+	queryTemplate := `"message.sender='%s'"`
+	return fmt.Sprintf(queryTemplate, address)
+}
+
+func buildRecvdTxsQuery(address string) string {
+	queryTemplate := `"transfer.recipient='%s'"`
+	return fmt.Sprintf(queryTemplate, address)
+}
+
+// given 2 sets of tx history, remove duplicates from the second set (recipient txs)
+func uniqueify(sender []client.TxSearchResponseResultTxs, recip []client.TxSearchResponseResultTxs) ([]client.TxSearchResponseResultTxs, []client.TxSearchResponseResultTxs) {
+	m := make(map[string]bool, len(sender)+len(recip))
+	filteredRecip := make([]client.TxSearchResponseResultTxs, 0, len(recip))
+	for i := 0; i < len(sender); i++ {
+		tx := sender[i]
+		m[*tx.Hash] = true
+	}
+	for i := 0; i < len(recip); i++ {
+		tx := recip[i]
+		if _, ok := m[*tx.Hash]; !ok {
+			m[*tx.Hash] = true
+			filteredRecip = append(filteredRecip, tx)
+		}
+	}
+	return sender, filteredRecip
+}
+
+func (c *Cursor) encode() *string {
```

Comment:
I think it makes more sense to return `(string, error)` and just catch the error if unable to encode. Able to remove `logger.Errorf`, `else` and extra var for returning pointer.

Reply:
sounds reasonable